### PR TITLE
Evict blocks based on size

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

With sustained large blocks (like observed with the hypersdk), the existing block caches evict too slowly. This changes the eviction policy to be size based.

## How this works

Switch the usage of `cache.LRU` to `cache.SizedLRU`

## How this was tested

- [X] CI
- [X] HyperSDK CI
- [ ] Fuji Canary
- [ ] Mainnet Canary

### CODE REVIEW

**Comment by patrick-ogrady** [comment]
This resolved this issue: https://github.com/ava-labs/hypersdk/issues/303

File: snow/engine/snowman/transitive.go [hunk 1]
```diff
@@ -24,10 +24,14 @@ import (
 	"github.com/ava-labs/avalanchego/utils/bag"
 	"github.com/ava-labs/avalanchego/utils/logging"
 	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/utils/units"
 	"github.com/ava-labs/avalanchego/utils/wrappers"
 )
 
-const nonVerifiedCacheSize = 128
+const (
+	nonVerifiedCacheSize = 128 * units.MiB
```
Comment (abi87) [inline snow/engine/snowman/transitive.go:32]: @StephenButtolph just to very I understand the issue here:
These hyper blocks are beefy and are quickly verified.
With item based cash they are kept in memory until other 128 blocks arrive, which is (much) more than hyperSDK need to verify them.
A size-based cache would saturate faster and keep them less in memory (still longer than needed to verify them).
Is this correct?

File: vms/components/chain/state.go [hunk 2]
```diff
@@ -138,11 +141,21 @@ func (s *State) initialize(config *Config) {
...
+			},
+		),
+		missingBlocks: &cache.LRU[ids.ID, struct{}]{Size: config.MissingCacheSize},
+		unverifiedBlocks: cache.NewSizedLRU[ids.ID, *BlockWrapper](
+			config.UnverifiedCacheSize,
+			func(bw *BlockWrapper) int {
+				return len(bw.Bytes()) + 2*pointerOverhead
+			},
+		),
+		bytesToIDCache: &cache.LRU[string, ids.ID]{Size: config.BytesToIDCacheSize},
```
Comment (abi87) [inline vms/components/chain/state.go:158]: while we keep this cache item based, should we reduce its size? I don't expect it to make a big difference if we don't, but IIUC it'll hold infos much longer than expected now, given other caches are size based?

**Review by abi87 — COMMENTED** [review]
just a few questions to make sure I understand the change

File: snow/engine/snowman/transitive.go (hunk 1 above)
Comment (StephenButtolph) [reply snow/engine/snowman/transitive.go:32]: correct. This cache specifically probably didn't have a problem... But some of the other caches (with thousands of blocks) were causing issues.

File: vms/components/chain/state.go (hunk 2 above)
Comment (StephenButtolph) [reply vms/components/chain/state.go:158]: Great 👁️. I had forgotten that we had block bytes as a key in a cache.

File: vms/rpcchainvm/vm_client.go [hunk 3]
```diff
@@ -63,10 +64,10 @@ import (
 )
 
 const (
```
Comment (patrick-ogrady) [inline vms/rpcchainvm/vm_client.go:66]: Is it too verbose to add a TODO that these should be configs?

File: vms/components/chain/state.go [hunk 4]
```diff
@@ -17,8 +17,19 @@ import (
 	"github.com/ava-labs/avalanchego/snow/choices"
 	"github.com/ava-labs/avalanchego/snow/consensus/snowman"
 	"github.com/ava-labs/avalanchego/snow/engine/snowman/block"
+	"github.com/ava-labs/avalanchego/utils/wrappers"
 )
 
+const pointerOverhead = wrappers.LongLen
```
Comment (patrick-ogrady) [inline vms/components/chain/state.go:23]: Should we make this a shared const somewhere?

File: vms/components/chain/state.go [hunk 5]
```diff
@@ -17,8 +17,17 @@ import (
 	"github.com/ava-labs/avalanchego/snow/choices"
 	"github.com/ava-labs/avalanchego/snow/consensus/snowman"
 	"github.com/ava-labs/avalanchego/snow/engine/snowman/block"
+	"github.com/ava-labs/avalanchego/utils/constants"
 )
 
+func cachedBlockSize(_ ids.ID, bw *BlockWrapper) int {
+	return ids.IDLen + len(bw.Bytes()) + 2*constants.PointerOverhead
```
Comment (patrick-ogrady) [inline vms/components/chain/state.go:24]: Why is this 2x? 1x for inner block pointer and one for the pointer to the wrapper object?

**Review by patrick-ogrady — COMMENTED** [review]
Just one clarifying question

File: vms/components/chain/state.go (hunk 5 above)
Comment (StephenButtolph) [reply vms/components/chain/state.go:24]: > 1x for inner block pointer and one for the pointer to the wrapper object?
Correct - went through and made sure I was consistent (there was 1 case a tx wrapper wasn't being accounted for correctly)

Comment (patrick-ogrady) [reply vms/components/chain/state.go:24]: May make sense to comment this but that's def a nit

**Review by patrick-ogrady — APPROVED** [review]

**Pull request merged**